File: lintcode/reverse_linked_listII.py
"""
Definition of ListNode

class ListNode(object):

    def __init__(self, val, next=None):
        self.val = val
        self.next = next
"""
class Solution:
    """
    @param head: The head of linked list
    @param m: start position
    @param n: end position
    """
    def reverseBetween(self, head, m, n):
        dummy = ListNode(None)
        dummy.next, before = head, dummy
        for _ in xrange(m - 1):
            before = before.next
        begin, after = before.next, before.next.next
        for _ in xrange(n - m):
            begin.next, after.next, before.next = after.next, before.next, after
            after = begin.next
        return dummy.next

File: leetcode/reverse_linked_listII.js
/* Reverse a linked list from position m to n. Do it in-place and in one-pass.
 * Time Complexity: O(n)
 * Space Complexity: O(1)
 */

/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */
/**
 * @param {ListNode} head
 * @param {number} m
 * @param {number} n
 * @return {ListNode}
 */
var reverseBetween = function(head, m, n) {
    var dummy = new ListNode(null), before = dummy;
    dummy.next = head;
    for (var i = 0; i < m - 1; ++i)
        before = before.next;
    var begin = before.next, after = before.next.next;
    for (i = 0; i < n - m; ++i) {
        begin.next = after.next;
        after.next = before.next;
        before.next = after;
        after = begin.next;
    }
    return dummy.next;
};
